### ISSUE TITLE
YCSB table creation never ends on Cloud Spanner

### ISSUE THREAD

**Issue by sirafantas** [cmu-db/benchbase#91]
I am trying to benchmark Cloud Spanner with YCSB, however, YCSB table creation never ends.

```console
$ java -jar benchbase.jar -b ycsb -c config/spanner/ycsb_config.xml --create=true --load=true --execute=true
[DEBUG] 2021-12-17 05:56:26,426 [main]  com.oltpbenchmark.api.StatementDialects getSQLDialectPath - No dialect file in benchmarks/ycsb/dialect-spanner.xml
[DEBUG] 2021-12-17 05:56:26,428 [main]  com.oltpbenchmark.api.StatementDialects load - SKIP - No SQL dialect file was given.
[INFO ] 2021-12-17 05:56:26,440 [main]  com.oltpbenchmark.DBWorkload main - ======================================================================

Benchmark:     YCSB {com.oltpbenchmark.benchmarks.ycsb.YCSBBenchmark}
Configuration: config/spanner/ycsb_config.xml
Type:          SPANNER
Driver:        com.google.cloud.spanner.jdbc.JdbcDriver
URL:           jdbc:cloudspanner:/projects/foo/instances/bar/databases/sira?credentials=/Users/sira/Code/benchbase/config/spanner/sira-personal-project-48ff2575be0b.json
Isolation:     TRANSACTION_SERIALIZABLE
Batch Size:    128
Scale Factor:  1.0
Terminals:     1

[INFO ] 2021-12-17 05:56:26,440 [main]  com.oltpbenchmark.DBWorkload main - ======================================================================
[DEBUG] 2021-12-17 05:56:26,464 [main]  com.oltpbenchmark.DBWorkload main - Using the following transaction types: [com.oltpbenchmark.api.TransactionType$Invalid/00, com.oltpbenchmark.benchmarks.ycsb.procedures.ReadRecord/01, com.oltpbenchmark.benchmarks.ycsb.procedures.InsertRecord/02, com.oltpbenchmark.benchmarks.ycsb.procedures.ScanRecord/03, com.oltpbenchmark.benchmarks.ycsb.procedures.UpdateRecord/04, com.oltpbenchmark.benchmarks.ycsb.procedures.DeleteRecord/05, com.oltpbenchmark.benchmarks.ycsb.procedures.ReadModifyWriteRecord/06]
[DEBUG] 2021-12-17 05:56:26,465 [main]  com.oltpbenchmark.DBWorkload main - Num groupings: 0
[DEBUG] 2021-12-17 05:56:26,474 [main]  com.oltpbenchmark.DBWorkload isBooleanOptionSet - CommandLine has option 'create'. Checking whether set to true
[DEBUG] 2021-12-17 05:56:26,474 [main]  com.oltpbenchmark.DBWorkload isBooleanOptionSet - CommandLine create => true
[INFO ] 2021-12-17 05:56:26,474 [main]  com.oltpbenchmark.DBWorkload main - Creating new YCSB database...
[DEBUG] 2021-12-17 05:56:26,474 [main]  com.oltpbenchmark.DBWorkload runCreator - Creating ycsb Database
[DEBUG] 2021-12-17 05:56:27,802 [main]  com.oltpbenchmark.api.BenchmarkModule createDatabase - Executing script [benchmarks/ycsb/ddl-spanner.sql] for database type [SPANNER]
[DEBUG] 2021-12-17 05:56:27,802 [main]  com.oltpbenchmark.util.ScriptRunner runScript - trying to find file by path benchmarks/ycsb/ddl-spanner.sql
[DEBUG] 2021-12-17 05:56:27,804 [main]  com.oltpbenchmark.util.ScriptRunner runScript - CREATE TABLE usertable (
[DEBUG] 2021-12-17 05:56:27,804 [main]  com.oltpbenchmark.util.ScriptRunner runScript -     ycsb_key INT64,
[DEBUG] 2021-12-17 05:56:27,804 [main]  com.oltpbenchmark.util.ScriptRunner runScript -     field1   STRING(100),
[DEBUG] 2021-12-17 05:56:27,804 [main]  com.oltpbenchmark.util.ScriptRunner runScript -     field2   STRING(100),
[DEBUG] 2021-12-17 05:56:27,805 [main]  com.oltpbenchmark.util.ScriptRunner runScript -     field3   STRING(100),
[DEBUG] 2021-12-17 05:56:27,805 [main]  com.oltpbenchmark.util.ScriptRunner runScript -     field4   STRING(100),
[DEBUG] 2021-12-17 05:56:27,806 [main]  com.oltpbenchmark.util.ScriptRunner runScript -     field5   STRING(100),
[DEBUG] 2021-12-17 05:56:27,806 [main]  com.oltpbenchmark.util.ScriptRunner runScript -     field6   STRING(100),
[DEBUG] 2021-12-17 05:56:27,806 [main]  com.oltpbenchmark.util.ScriptRunner runScript -     field7   STRING(100),
[DEBUG] 2021-12-17 05:56:27,806 [main]  com.oltpbenchmark.util.ScriptRunner runScript -     field8   STRING(100),
[DEBUG] 2021-12-17 05:56:27,806 [main]  com.oltpbenchmark.util.ScriptRunner runScript -     field9   STRING(100),
[DEBUG] 2021-12-17 05:56:27,806 [main]  com.oltpbenchmark.util.ScriptRunner runScript -     field10  STRING(100),
[DEBUG] 2021-12-17 05:56:27,806 [main]  com.oltpbenchmark.util.ScriptRunner runScript - ) PRIMARY KEY (ycsb_key);

(never ends...)
```

I have already checked that the credentials work well and the same table can be created using simple code like [this](https://github.com/googleapis/java-spanner-jdbc#creating-a-jdbc-connection).

I couldn't make any progress on this problem though I investigated the source code.

**Comment by timveil** [comment]
@y-sira I was not able to recreate your issue.  I just tested with the Spanner Emulator and was able to successfully run `ycsb` against the emulator following the steps here... https://github.com/cmu-db/benchbase/tree/main/docker/spanner-emulator

Have you had a chance to try that?

**Comment by timveil** [comment]
One thing i am noticing is that there may be a problem with the latest version of the spanner jdbc driver.  We recently merged in the latest version `2.5.5`.  I'm running into problems testing against the emulator with this version.  If i revert the jdbc driver back to version `2.5.1` ycsb executes fine.  If you are running the lasted code from the `main` branch you may want to change the `pom.xml` file to point to while more research is done.
```
<dependency>
    <groupId>com.google.cloud</groupId>
    <artifactId>google-cloud-spanner-jdbc</artifactId>
    <version>2.5.1</version>
</dependency>
```

**Comment by y-sira** [comment]
>  If i revert the jdbc driver back to version `2.5.1` ycsb executes fine.

It works for me too.

However, the following `load` step did not work.
It seems to be a Spanner-side problem.

```
$ java -jar benchbase.jar -b ycsb -c config/spanner/ycsb_config.xml --create=true --load=true --execute=true
[DEBUG] 2021-12-17 11:53:25,000 [main]  com.oltpbenchmark.api.StatementDialects getSQLDialectPath - No dialect file in benchmarks/ycsb/dialect-spanner.xml
[DEBUG] 2021-12-17 11:53:25,003 [main]  com.oltpbenchmark.api.StatementDialects load - SKIP - No SQL dialect file was given.
[INFO ] 2021-12-17 11:53:25,020 [main]  com.oltpbenchmark.DBWorkload main - ======================================================================

Benchmark:     YCSB {com.oltpbenchmark.benchmarks.ycsb.YCSBBenchmark}
Configuration: config/spanner/ycsb_config.xml
Type:          SPANNER
Driver:        com.google.cloud.spanner.jdbc.JdbcDriver
URL:           jdbc:cloudspanner:/projects/foo/instances/bar/databases/sira?credentials=/Users/sira/Code/benchbase/config/spanner/sira-personal-project-48ff2575be0b.json
Isolation:     TRANSACTION_SERIALIZABLE
Batch Size:    128
Scale Factor:  1.0
Terminals:     1

[INFO ] 2021-12-17 11:53:25,020 [main]  com.oltpbenchmark.DBWorkload main - ======================================================================
[DEBUG] 2021-12-17 11:53:25,049 [main]  com.oltpbenchmark.DBWorkload main - Using the following transaction types: [com.oltpbenchmark.api.TransactionType$Invalid/00, com.oltpbenchmark.benchmarks.ycsb.procedures.ReadRecord/01, com.oltpbenchmark.benchmarks.ycsb.procedures.InsertRecord/02, com.oltpbenchmark.benchmarks.ycsb.procedures.ScanRecord/03, com.oltpbenchmark.benchmarks.ycsb.procedures.UpdateRecord/04, com.oltpbenchmark.benchmarks.ycsb.procedures.DeleteRecord/05, com.oltpbenchmark.benchmarks.ycsb.procedures.ReadModifyWriteRecord/06]
[DEBUG] 2021-12-17 11:53:25,051 [main]  com.oltpbenchmark.DBWorkload main - Num groupings: 0
[DEBUG] 2021-12-17 11:53:25,065 [main]  com.oltpbenchmark.DBWorkload isBooleanOptionSet - CommandLine has option 'create'. Checking whether set to true
[DEBUG] 2021-12-17 11:53:25,065 [main]  com.oltpbenchmark.DBWorkload isBooleanOptionSet - CommandLine create => true
[INFO ] 2021-12-17 11:53:25,065 [main]  com.oltpbenchmark.DBWorkload main - Creating new YCSB database...
[DEBUG] 2021-12-17 11:53:25,066 [main]  com.oltpbenchmark.DBWorkload runCreator - Creating ycsb Database
[DEBUG] 2021-12-17 11:53:27,015 [main]  com.oltpbenchmark.api.BenchmarkModule createDatabase - Executing script [benchmarks/ycsb/ddl-spanner.sql] for database type [SPANNER]
[DEBUG] 2021-12-17 11:53:27,015 [main]  com.oltpbenchmark.util.ScriptRunner runScript - trying to find file by path benchmarks/ycsb/ddl-spanner.sql
[DEBUG] 2021-12-17 11:53:27,017 [main]  com.oltpbenchmark.util.ScriptRunner runScript - CREATE TABLE usertable (
[DEBUG] 2021-12-17 11:53:27,017 [main]  com.oltpbenchmark.util.ScriptRunner runScript -     ycsb_key INT64,
[DEBUG] 2021-12-17 11:53:27,017 [main]  com.oltpbenchmark.util.ScriptRunner runScript -     field1   STRING(100),
[DEBUG] 2021-12-17 11:53:27,017 [main]  com.oltpbenchmark.util.ScriptRunner runScript -     field2   STRING(100),
[DEBUG] 2021-12-17 11:53:27,019 [main]  com.oltpbenchmark.util.ScriptRunner runScript -     field3   STRING(100),
[DEBUG] 2021-12-17 11:53:27,019 [main]  com.oltpbenchmark.util.ScriptRunner runScript -     field4   STRING(100),
[DEBUG] 2021-12-17 11:53:27,019 [main]  com.oltpbenchmark.util.ScriptRunner runScript -     field5   STRING(100),
[DEBUG] 2021-12-17 11:53:27,019 [main]  com.oltpbenchmark.util.ScriptRunner runScript -     field6   STRING(100),
[DEBUG] 2021-12-17 11:53:27,019 [main]  com.oltpbenchmark.util.ScriptRunner runScript -     field7   STRING(100),
[DEBUG] 2021-12-17 11:53:27,019 [main]  com.oltpbenchmark.util.ScriptRunner runScript -     field8   STRING(100),
[DEBUG] 2021-12-17 11:53:27,020 [main]  com.oltpbenchmark.util.ScriptRunner runScript -     field9   STRING(100),
[DEBUG] 2021-12-17 11:53:27,020 [main]  com.oltpbenchmark.util.ScriptRunner runScript -     field10  STRING(100),
[DEBUG] 2021-12-17 11:53:27,020 [main]  com.oltpbenchmark.util.ScriptRunner runScript - ) PRIMARY KEY (ycsb_key);
[INFO ] 2021-12-17 11:53:51,982 [main]  com.oltpbenchmark.DBWorkload main - Finished creating new YCSB database...
[DEBUG] 2021-12-17 11:53:53,385 [main]  com.oltpbenchmark.DBWorkload main - Skipping clearing benchmark database tables
[DEBUG] 2021-12-17 11:53:53,385 [main]  com.oltpbenchmark.DBWorkload isBooleanOptionSet - CommandLine has option 'load'. Checking whether set to true
[DEBUG] 2021-12-17 11:53:53,385 [main]  com.oltpbenchmark.DBWorkload isBooleanOptionSet - CommandLine load => true
[INFO ] 2021-12-17 11:53:53,385 [main]  com.oltpbenchmark.DBWorkload main - Loading data into YCSB database...
[DEBUG] 2021-12-17 11:53:53,385 [main]  com.oltpbenchmark.DBWorkload runLoader - Loading ycsb Database
[DEBUG] 2021-12-17 11:53:53,388 [main]  com.oltpbenchmark.api.Loader <init> - # of RECORDS:  1000
[INFO ] 2021-12-17 11:53:53,389 [main]  com.oltpbenchmark.util.ThreadUtil runLoaderThreads - Creating a Thread Pool with a size of 1 to run 1 Loader Threads.  0 threads will be queued.
[DEBUG] 2021-12-17 11:53:53,393 [Thread-2]  com.oltpbenchmark.api.Loader load - YCSBLoadThread[0, 1000]
[ERROR] 2021-12-17 11:54:23,477 [Thread-2]  com.oltpbenchmark.api.LoaderThread run - Unexpected error when loading YCSB database
[ERROR] 2021-12-17 11:54:23,479 [Thread-2]  com.oltpbenchmark.util.ThreadUtil run - Exception in Loader Thread with message: [com.google.cloud.spanner.jdbc.JdbcSqlExceptionFactory$JdbcSqlTimeoutException: DEADLINE_EXCEEDED: com.google.api.gax.rpc.DeadlineExceededException: io.grpc.StatusRuntimeException: DEADLINE_EXCEEDED: deadline exceeded after 29.996471577s. [closed=[], open=[[remote_addr=spanner.googleapis.com/142.251.42.138:443]]]]; will count down latch with count 1 and then exit :(
java.lang.RuntimeException: com.google.cloud.spanner.jdbc.JdbcSqlExceptionFactory$JdbcSqlTimeoutException: DEADLINE_EXCEEDED: com.google.api.gax.rpc.DeadlineExceededException: io.grpc.StatusRuntimeException: DEADLINE_EXCEEDED: deadline exceeded after 29.996471577s. [closed=[], open=[[remote_addr=spanner.googleapis.com/142.251.42.138:443]]]
	at com.oltpbenchmark.api.LoaderThread.run(LoaderThread.java:50)
	at com.oltpbenchmark.util.ThreadUtil$LatchRunnable.run(ThreadUtil.java:145)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: com.google.cloud.spanner.jdbc.JdbcSqlExceptionFactory$JdbcSqlTimeoutException: DEADLINE_EXCEEDED: com.google.api.gax.rpc.DeadlineExceededException: io.grpc.StatusRuntimeException: DEADLINE_EXCEEDED: deadline exceeded after 29.996471577s. [closed=[], open=[[remote_addr=spanner.googleapis.com/142.251.42.138:443]]]
	at com.google.cloud.spanner.jdbc.JdbcSqlExceptionFactory.of(JdbcSqlExceptionFactory.java:205)
	at com.google.cloud.spanner.jdbc.JdbcStatement.executeBatch(JdbcStatement.java:283)
	at com.google.cloud.spanner.jdbc.JdbcStatement.executeBatch(JdbcStatement.java:260)
	at com.oltpbenchmark.benchmarks.ycsb.YCSBLoader.loadRecords(YCSBLoader.java:80)
	at com.oltpbenchmark.benchmarks.ycsb.YCSBLoader$1.load(YCSBLoader.java:56)
	at com.oltpbenchmark.api.LoaderThread.run(LoaderThread.java:45)
	... 4 more
Caused by: com.google.cloud.spanner.SpannerException: DEADLINE_EXCEEDED: com.google.api.gax.rpc.DeadlineExceededException: io.grpc.StatusRuntimeException: DEADLINE_EXCEEDED: deadline exceeded after 29.996471577s. [closed=[], open=[[remote_addr=spanner.googleapis.com/142.251.42.138:443]]]
	at com.google.cloud.spanner.SpannerExceptionFactory.newSpannerExceptionPreformatted(SpannerExceptionFactory.java:284)
	at com.google.cloud.spanner.SpannerExceptionFactory.newSpannerException(SpannerExceptionFactory.java:61)
	at com.google.cloud.spanner.SpannerExceptionFactory.newSpannerException(SpannerExceptionFactory.java:181)
	at com.google.cloud.spanner.spi.v1.GapicSpannerRpc.get(GapicSpannerRpc.java:1676)
	at com.google.cloud.spanner.spi.v1.GapicSpannerRpc.executeBatchDml(GapicSpannerRpc.java:1520)
	at com.google.cloud.spanner.TransactionRunnerImpl$TransactionContextImpl.batchUpdate(TransactionRunnerImpl.java:785)
	at com.google.cloud.spanner.connection.SingleUseTransaction.lambda$null$5(SingleUseTransaction.java:388)
	at com.google.cloud.spanner.TransactionRunnerImpl.lambda$runInternal$0(TransactionRunnerImpl.java:960)
	at com.google.api.gax.retrying.DirectRetryingExecutor.submit(DirectRetryingExecutor.java:105)
	at com.google.cloud.RetryHelper.run(RetryHelper.java:76)
	at com.google.cloud.RetryHelper.runWithRetries(RetryHelper.java:50)
	at com.google.cloud.spanner.SpannerRetryHelper.runTxWithRetriesOnAborted(SpannerRetryHelper.java:79)
	at com.google.cloud.spanner.SpannerRetryHelper.runTxWithRetriesOnAborted(SpannerRetryHelper.java:68)
	at com.google.cloud.spanner.TransactionRunnerImpl.runInternal(TransactionRunnerImpl.java:1017)
	at com.google.cloud.spanner.TransactionRunnerImpl.run(TransactionRunnerImpl.java:922)
	at com.google.cloud.spanner.SessionPool$SessionPoolTransactionRunner.run(SessionPool.java:934)
	at com.google.cloud.spanner.connection.SingleUseTransaction.lambda$executeTransactionalBatchUpdateAsync$6(SingleUseTransaction.java:384)
	at io.grpc.Context$2.call(Context.java:596)
	at com.google.common.util.concurrent.TrustedListenableFutureTask$TrustedFutureInterruptibleTask.runInterruptibly(TrustedListenableFutureTask.java:131)
	at com.google.common.util.concurrent.InterruptibleTask.run(InterruptibleTask.java:74)
	at com.google.common.util.concurrent.TrustedListenableFutureTask.run(TrustedListenableFutureTask.java:82)
	... 3 more
	Suppressed: com.google.cloud.spanner.connection.AbstractBaseUnitOfWork$SpannerAsyncExecutionException: Execution failed for statement: RUN BATCH
		at com.google.cloud.spanner.connection.AbstractBaseUnitOfWork.executeStatementAsync(AbstractBaseUnitOfWork.java:233)
		at com.google.cloud.spanner.connection.AbstractBaseUnitOfWork.executeStatementAsync(AbstractBaseUnitOfWork.java:145)
		at com.google.cloud.spanner.connection.SingleUseTransaction.executeTransactionalBatchUpdateAsync(SingleUseTransaction.java:403)
		at com.google.cloud.spanner.connection.SingleUseTransaction.executeBatchUpdateAsync(SingleUseTransaction.java:325)
		at com.google.cloud.spanner.connection.ConnectionImpl.internalExecuteBatchUpdateAsync(ConnectionImpl.java:1070)
		at com.google.cloud.spanner.connection.ConnectionImpl.executeBatchUpdate(ConnectionImpl.java:978)
		at com.google.cloud.spanner.jdbc.JdbcStatement.executeBatch(JdbcStatement.java:274)
		at com.google.cloud.spanner.jdbc.JdbcStatement.executeBatch(JdbcStatement.java:260)
		at com.oltpbenchmark.benchmarks.ycsb.YCSBLoader.loadRecords(YCSBLoader.java:80)
		at com.oltpbenchmark.benchmarks.ycsb.YCSBLoader$1.load(YCSBLoader.java:56)
		at com.oltpbenchmark.api.LoaderThread.run(LoaderThread.java:45)
		at com.oltpbenchmark.util.ThreadUtil$LatchRunnable.run(ThreadUtil.java:145)
		... 3 more
Caused by: java.util.concurrent.ExecutionException: com.google.api.gax.rpc.DeadlineExceededException: io.grpc.StatusRuntimeException: DEADLINE_EXCEEDED: deadline exceeded after 29.996471577s. [closed=[], open=[[remote_addr=spanner.googleapis.com/142.251.42.138:443]]]
	at com.google.common.util.concurrent.AbstractFuture.getDoneValue(AbstractFuture.java:588)
	at com.google.common.util.concurrent.AbstractFuture.get(AbstractFuture.java:567)
	at com.google.cloud.spanner.spi.v1.GapicSpannerRpc.get(GapicSpannerRpc.java:1670)
	... 20 more
Caused by: com.google.api.gax.rpc.DeadlineExceededException: io.grpc.StatusRuntimeException: DEADLINE_EXCEEDED: deadline exceeded after 29.996471577s. [closed=[], open=[[remote_addr=spanner.googleapis.com/142.251.42.138:443]]]
	at com.google.api.gax.rpc.ApiExceptionFactory.createException(ApiExceptionFactory.java:51)
	at com.google.api.gax.grpc.GrpcApiExceptionFactory.create(GrpcApiExceptionFactory.java:72)
	at com.google.api.gax.grpc.GrpcApiExceptionFactory.create(GrpcApiExceptionFactory.java:60)
	at com.google.api.gax.grpc.GrpcExceptionCallable$ExceptionTransformingFuture.onFailure(GrpcExceptionCallable.java:97)
	at com.google.api.core.ApiFutures$1.onFailure(ApiFutures.java:68)
	at com.google.common.util.concurrent.Futures$CallbackListener.run(Futures.java:1133)
	at com.google.common.util.concurrent.DirectExecutor.execute(DirectExecutor.java:31)
	at com.google.common.util.concurrent.AbstractFuture.executeListener(AbstractFuture.java:1277)
	at com.google.common.util.concurrent.AbstractFuture.complete(AbstractFuture.java:1038)
	at com.google.common.util.concurrent.AbstractFuture.setException(AbstractFuture.java:808)
	at io.grpc.stub.ClientCalls$GrpcFuture.setException(ClientCalls.java:563)
	at io.grpc.stub.ClientCalls$UnaryStreamToFuture.onClose(ClientCalls.java:533)
	at io.grpc.PartialForwardingClientCallListener.onClose(PartialForwardingClientCallListener.java:39)
	at io.grpc.ForwardingClientCallListener.onClose(ForwardingClientCallListener.java:23)
	at io.grpc.ForwardingClientCallListener$SimpleForwardingClientCallListener.onClose(ForwardingClientCallListener.java:40)
	at io.grpc.PartialForwardingClientCallListener.onClose(PartialForwardingClientCallListener.java:39)
	at io.grpc.ForwardingClientCallListener.onClose(ForwardingClientCallListener.java:23)
	at io.grpc.ForwardingClientCallListener$SimpleForwardingClientCallListener.onClose(ForwardingClientCallListener.java:40)
	at com.google.cloud.spanner.spi.v1.SpannerErrorInterceptor$1$1.onClose(SpannerErrorInterceptor.java:100)
	at io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:557)
	at io.grpc.internal.ClientCallImpl.access$300(ClientCallImpl.java:69)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInternal(ClientCallImpl.java:738)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext(ClientCallImpl.java:717)
	at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
	at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:133)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
	... 3 more
Caused by: io.grpc.StatusRuntimeException: DEADLINE_EXCEEDED: deadline exceeded after 29.996471577s. [closed=[], open=[[remote_addr=spanner.googleapis.com/142.251.42.138:443]]]
	at io.grpc.Status.asRuntimeException(Status.java:535)
	... 20 more
```

**Comment by y-sira** [comment]
`load` works after reducing `batchsize` to 64 from 128. 
I will share my experience with you after further inspection.

**Comment by y-sira** [comment]
FYI:

```console
$ cat config/spanner/ycsb_config.xml 
<?xml version="1.0"?>
<parameters>
    <!-- Connection details -->
    <type>SPANNER</type>
    <driver>com.google.cloud.spanner.jdbc.JdbcDriver</driver>
    <url>jdbc:cloudspanner:/projects/foo/instances/bar/databases/sira?credentials=/Users/sira/Code/benchbase/config/spanner/sira-personal-project-48ff2575be0b.json</url>
    <isolation>TRANSACTION_SERIALIZABLE</isolation>
    <batchsize>32</batchsize>

    <!-- Scalefactor in YCSB is *1000 the number of rows in the USERTABLE-->
    <scalefactor>100</scalefactor>

    <!-- The workload -->
    <terminals>100</terminals>
    <works>
        <work>
            <warmup>60</warmup>
            <time>60</time>
            <rate>unlimited</rate>
            <weights>50,5,15,10,10,10</weights>
        </work>
    </works>

    <!-- YCSB Procedures declaration -->
    <transactiontypes>
        <transactiontype>
            <name>ReadRecord</name>
        </transactiontype>
        <transactiontype>
            <name>InsertRecord</name>
        </transactiontype>
        <transactiontype>
            <name>ScanRecord</name>
        </transactiontype>
        <transactiontype>
            <name>UpdateRecord</name>
        </transactiontype>
        <transactiontype>
            <name>DeleteRecord</name>
        </transactiontype>
        <transactiontype>
            <name>ReadModifyWriteRecord</name>
        </transactiontype>
    </transactiontypes>
</parameters>
```

```
$ cat results/ycsb_2021-12-19_01-36-03.summary.json
{
 "scalefactor": "100",
 "Current Timestamp (milliseconds)": 1639845363482,
 "Benchmark Type": "ycsb",
 "isolation": "TRANSACTION_SERIALIZABLE",
 "DBMS Version": null,
 "terminals": "100",
 "DBMS Type": "SPANNER",
 "Latency Distribution": {
  "95th Percentile Latency (microseconds)": 562434,
  "Maximum Latency (microseconds)": 5096529,
  "Median Latency (microseconds)": 291064,
  "Minimum Latency (microseconds)": 271825,
  "25th Percentile Latency (microseconds)": 283227,
  "90th Percentile Latency (microseconds)": 508923,
  "99th Percentile Latency (microseconds)": 882095,
  "75th Percentile Latency (microseconds)": 354341,
  "Average Latency (microseconds)": 353334
 },
 "Throughput (requests/second)": 280.9889253366542
}
```

```
$ cat results/ycsb_2021-12-19_01-28-27.summary.json 
{
 "scalefactor": "100",
 "Current Timestamp (milliseconds)": 1639844908672,
 "Benchmark Type": "ycsb",
 "isolation": "TRANSACTION_SERIALIZABLE",
 "DBMS Version": null,
 "terminals": "1000",
 "DBMS Type": "SPANNER",
 "Latency Distribution": {
  "95th Percentile Latency (microseconds)": 1180391,
  "Maximum Latency (microseconds)": 2921139,
  "Median Latency (microseconds)": 919318,
  "Minimum Latency (microseconds)": 775136,
  "25th Percentile Latency (microseconds)": 875734,
  "90th Percentile Latency (microseconds)": 1120674,
  "99th Percentile Latency (microseconds)": 1299963,
  "75th Percentile Latency (microseconds)": 1005675,
  "Average Latency (microseconds)": 955260
 },
 "Throughput (requests/second)": 1029.9674214889828
}
```

I got it why many companies create their own distributed transactional database...

**Comment by timveil** [comment]
this issue will ultimately be resolved when #97 is closed
